<h2>Optimización de Redes Neuronales para la Clasificación de Actividad Muscular</h2>
<p>En este ejercicio trabajarás con datos que vienen de un experimento en el que se midió actividad muscular con la técnica de la Electromiografía en el brazo derecho de varios participantes cuando éstos realizaban un movimiento con la mano entre siete posibles (Flexionar hacia arriba, Flexionar hacia abajo, Cerrar la mano, Estirar la mano, Abrir la mano, Coger un objeto, No moverse). Al igual que en el ejercicio anterior, los datos se cargan con la función <code>loadtxt</code> de <a href="https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html" target="_blank">numpy</a>. A su vez, la primera columna corresponde a la clase (1, 2, 3, 4, 5, 6, y 7), la segunda columna se ignora, y el resto de las columnas indican las variables que se calcularon de la respuesta muscular. El archivo de datos con el que trabajarás depende de tu matrícula.</p>

<p>Para este conjunto de datos:</p>

<ol>
  <li>Evalúa un modelo perceptrón multicapa con validación cruzada utilizando al menos 5 capas de 20 neuronas.</li>
  <li>Evalúa un modelo perceptrón multicapa con validación cruzada, pero encontrando el número óptimo de capas y neuronas de la red.</li>
  <li>Prepara el modelo perceptrón multicapa:
    <ul>
      <li>Obtén los hiperparámetros óptimos de capas y neuronas de la red.</li>
      <li>Con los hiperparámetros óptimos, ajusta el modelo con todos los datos.</li>
    </ul>
  </li>
  <li>Contesta lo siguientes:
    <ul>
      <li>¿Observas alguna mejora importante al optimizar el tamaño de la red? ¿Es el resultado que esperabas? Argumenta tu respuesta.</li>
      <li>¿Qué inconvenientes hay al encontrar el tamaño óptimo de la red? ¿Por qué?</li>
    </ul>
  </li>
</ol>

<p><strong>Nota:</strong> El conjunto de datos con el que trabajarás en este ejercicio depende del penúltimo número de tu matrícula de acuerdo a la siguiente lista:</p>

<ul>
  <li><strong>0 y 1</strong> - <a href="link_M_1">M_1.txt</a></li>
  <li><strong>2 y 3</strong> - <a href="link_M_2">M_2.txt</a></li>
  <li><strong>4 y 5</strong> - <a href="link_M_3">M_3.txt</a></li>
  <li><strong>6 y 7</strong> - <a href="link_M_4">M_4.txt</a></li>
  <li><strong>8 y 9</strong> - <a href="link_M_5">M_5.txt</a></li>
</ul>

In [106]:
import numpy as np
import pandas as pd
from keras import Sequential
from sklearn.model_selection import KFold
from keras.layers import Dense
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.preprocessing import MinMaxScaler
from keras.utils import to_categorical
from sklearn.linear_model import LinearRegression
from sklearn.metrics import classification_report

In [107]:
dataframe=np.loadtxt('data/problem2/M_5.txt')
dataframe

array([[ 1.        ,  1.        ,  0.15991005, ...,  0.87855426,
         1.63603899,  1.60896884],
       [ 1.        ,  1.        , -1.03964581, ...,  0.67610431,
         0.75054922,  1.04006595],
       [ 1.        ,  1.        , -1.41164407, ...,  0.36203866,
         1.59177889,  1.53300746],
       ...,
       [ 7.        ,  1.        , -7.71798749, ..., -5.14926718,
        -2.02700281, -7.51078954],
       [ 7.        ,  1.        , -7.44747637, ..., -4.12011728,
        -1.65436398, -7.70478237],
       [ 7.        ,  1.        , -8.05939761, ..., -6.3178266 ,
        -2.71340576, -8.45893342]])

In [108]:
x=dataframe[:,2:]
x

array([[ 0.15991005,  0.82903782, -0.23632194, ...,  0.87855426,
         1.63603899,  1.60896884],
       [-1.03964581,  0.06158144, -0.37280361, ...,  0.67610431,
         0.75054922,  1.04006595],
       [-1.41164407, -1.09091459, -1.16421303, ...,  0.36203866,
         1.59177889,  1.53300746],
       ...,
       [-7.71798749, -5.5483518 , -4.39917235, ..., -5.14926718,
        -2.02700281, -7.51078954],
       [-7.44747637, -5.2238522 , -4.0738337 , ..., -4.12011728,
        -1.65436398, -7.70478237],
       [-8.05939761, -5.36202795, -4.4292454 , ..., -6.3178266 ,
        -2.71340576, -8.45893342]])

In [109]:
y=dataframe[:,0]
y

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3.

In [110]:
scaler=MinMaxScaler()
x=scaler.fit_transform(x)
x

array([[0.81598861, 0.92358857, 0.62635428, ..., 0.62266153, 0.92014708,
        0.84719293],
       [0.70645297, 0.82383374, 0.61035687, ..., 0.6051447 , 0.73532685,
        0.80072764],
       [0.67248451, 0.67403102, 0.51759347, ..., 0.5779704 , 0.91090908,
        0.84098876],
       ...,
       [0.09663024, 0.09464838, 0.13841448, ..., 0.10110874, 0.15559365,
        0.1023344 ],
       [0.12133156, 0.1368272 , 0.17654836, ..., 0.19015517, 0.23337117,
        0.08648999],
       [0.06545489, 0.11886696, 0.13488953, ..., 0.        , 0.012327  ,
        0.02489451]])

 <h3>1. Evalúa un modelo perceptrón multicapa con validación cruzada utilizando al menos 5 capas de 20 neuronas.</h3>

In [111]:
features = x.shape[1]
features

630

In [112]:
num_classes = 7
y = to_categorical(y - 1, num_classes=num_classes)  # Restar 1 si las clases están en el rango [1, 7]

# Definir el número de características
features = x.shape[1]

n_folds = 5
kfold = KFold(n_splits=n_folds, shuffle=True)
cv_y_test = []
cv_y_pred = []
it = 0

for train_index, test_index in kfold.split(x, y):
    it += 1
    print("it: ", it)
    
    x_train = x[train_index, :]
    y_train = y[train_index]
    
    x_test = x[test_index, :]
    y_test = y[test_index]
    
    # Fase de entrenamiento
    # Definir el modelo MLP
    mlp_model = Sequential()
    mlp_model.add(Dense(20, activation='relu', input_dim=features))
    mlp_model.add(Dense(20, activation='relu'))
    mlp_model.add(Dense(20, activation='relu'))
    mlp_model.add(Dense(20, activation='relu'))
    mlp_model.add(Dense(20, activation='relu'))
    mlp_model.add(Dense(num_classes, activation='softmax'))  # Capa de salida
    
    # Compilar el modelo
    mlp_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    
    # Entrenar el modelo
    mlp_model.fit(x_train, y_train, epochs=500, batch_size=32, verbose=0)
    
    # Predecir en los datos de prueba
    y_pred_proba = mlp_model.predict(x_test)
    y_pred = np.argmax(y_pred_proba, axis=1)
    y_test_labels = np.argmax(y_test, axis=1)
    
    # Agregar los valores verdaderos y predicciones
    cv_y_test.append(y_test_labels)
    cv_y_pred.append(y_pred)

# Convertir listas de arrays a un solo array para la evaluación final
cv_y_test = np.concatenate(cv_y_test)
cv_y_pred = np.concatenate(cv_y_pred)

# Imprimir el reporte de clasificación
print(classification_report(cv_y_test, cv_y_pred))

it:  1
4/4 [==============================] - 0s 2ms/step
it:  2
4/4 [==============================] - 0s 2ms/step
it:  3
4/4 [==============================] - 0s 1ms/step
it:  4
4/4 [==============================] - 0s 2ms/step
it:  5
4/4 [==============================] - 0s 1ms/step
              precision    recall  f1-score   support

           0       0.81      0.94      0.87        90
           1       0.57      0.60      0.58        90
           2       1.00      0.93      0.97        90
           3       0.87      0.84      0.86        90
           4       0.94      0.86      0.90        90
           5       0.71      0.70      0.70        90
           6       0.98      0.96      0.97        89

    accuracy                           0.83       629
   macro avg       0.84      0.83      0.84       629
weighted avg       0.84      0.83      0.83       629



<h3>2. Evalúa un modelo perceptrón multicapa con validación cruzada, pero encontrando el número óptimo de capas y neuronas de la red.</h3>

In [113]:
# Configuraciones a probar
layer_configs = [
    [50, 20],               # 2 capas con diferentes números de neuronas
    [20, 20, 20, 20, 20],  # 5 capas con 20 neuronas cada una
    [40, 30, 20],           # 3 capas con diferentes números de neuronas
    [30, 30, 30, 30]       # 4 capas con 30 neuronas cada una
]

n_folds = 5
best_config = None
best_accuracy = 0

for config in layer_configs:
    print(f"Evaluando configuración: {config}")
    cv_y_test = []
    cv_y_pred = []
    
    kfold = KFold(n_splits=n_folds, shuffle=True)
    
    for train_index, test_index in kfold.split(x, y):
        x_train = x[train_index, :]
        y_train = y[train_index]
        
        x_test = x[test_index, :]
        y_test = y[test_index]
        
        # Fase de entrenamiento
        # Definir el modelo MLP
        mlp_model = Sequential()
        mlp_model.add(Dense(config[0], activation='relu', input_dim=features))
        
        for units in config[1:]:
            mlp_model.add(Dense(units, activation='relu'))
        
        mlp_model.add(Dense(num_classes, activation='softmax'))  # Capa de salida
        
        # Compilar el modelo
        mlp_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
        
        # Entrenar el modelo
        mlp_model.fit(x_train, y_train, epochs=500, batch_size=32, verbose=0)
        
        # Predecir en los datos de prueba
        y_pred_proba = mlp_model.predict(x_test)
        y_pred = np.argmax(y_pred_proba, axis=1)
        y_test_labels = np.argmax(y_test, axis=1)
        
        # Agregar los valores verdaderos y predicciones
        cv_y_test.append(y_test_labels)
        cv_y_pred.append(y_pred)

    # Convertir listas de arrays a un solo array para la evaluación final
    cv_y_test = np.concatenate(cv_y_test)
    cv_y_pred = np.concatenate(cv_y_pred)
    
    # Evaluar el rendimiento
    accuracy = np.mean(cv_y_pred == cv_y_test)
    print(f"Precisión para configuración {config}: {accuracy}")

    # Guardar la mejor configuración
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_config = config

print(f"La mejor configuración es {best_config} con una precisión de {best_accuracy}")

Evaluando configuración: [50, 20]
4/4 [==============================] - 0s 1ms/step
Precisión para configuración [50, 20]: 0.904610492845787
Evaluando configuración: [20, 20, 20, 20, 20]
4/4 [==============================] - 0s 2ms/step
Precisión para configuración [20, 20, 20, 20, 20]: 0.8553259141494436
Evaluando configuración: [40, 30, 20]
4/4 [==============================] - 0s 1ms/step
Precisión para configuración [40, 30, 20]: 0.8807631160572337
Evaluando configuración: [30, 30, 30, 30]
4/4 [==============================] - 0s 2ms/step
Precisión para configuración [30, 30, 30, 30]: 0.8791732909379968
La mejor configuración es [50, 20] con una precisión de 0.904610492845787


<h3>3. Prepara el modelo perceptrón multicapa:</h3>
    <ul>
      <li>Obtén los hiperparámetros óptimos de capas y neuronas de la red.</li>
      <li>Con los hiperparámetros óptimos, ajusta el modelo con todos los datos.</li>
    </ul>

In [114]:
# Ajustar el modelo con todos los datos usando la mejor configuración
x_train = x
y_train = y

mlp_model = Sequential()
mlp_model.add(Dense(best_config[0], activation='relu', input_dim=features))

for units in best_config[1:]:
    mlp_model.add(Dense(units, activation='relu'))

mlp_model.add(Dense(num_classes, activation='softmax'))  # Capa de salida

# Compilar el modelo
mlp_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Entrenar el modelo
mlp_model.fit(x_train, y_train, epochs=500, batch_size=32, verbose=1)

# Predecir en todos los datos
y_pred_proba = mlp_model.predict(x_train)
y_pred = np.argmax(y_pred_proba, axis=1)
y_true = np.argmax(y_train, axis=1)

# Imprimir el reporte de clasificación
print(classification_report(y_true, y_pred))

Epoch 1/500
20/20 [==============================] - 1s 1ms/step - loss: 1.9472 - accuracy: 0.1526 
Epoch 2/500
20/20 [==============================] - 0s 1ms/step - loss: 1.8804 - accuracy: 0.1717
Epoch 3/500
20/20 [==============================] - 0s 1ms/step - loss: 1.7833 - accuracy: 0.2846
Epoch 4/500
20/20 [==============================] - 0s 2ms/step - loss: 1.6865 - accuracy: 0.4086
Epoch 5/500
20/20 [==============================] - 0s 2ms/step - loss: 1.6039 - accuracy: 0.4976
Epoch 6/500
20/20 [==============================] - 0s 2ms/step - loss: 1.4877 - accuracy: 0.5723
Epoch 7/500
20/20 [==============================] - 0s 2ms/step - loss: 1.3709 - accuracy: 0.6200
Epoch 8/500
20/20 [==============================] - 0s 1ms/step - loss: 1.2593 - accuracy: 0.6614
Epoch 9/500
20/20 [==============================] - 0s 2ms/step - loss: 1.1739 - accuracy: 0.6280
Epoch 10/500
20/20 [==============================] - 0s 1ms/step - loss: 1.1429 - accuracy: 0.6010
Epoch 11

<h3>4. Contesta lo siguientes:</h3>

<h4>4.1 ¿Observas alguna mejora importante al optimizar el tamaño de la red? ¿Es el resultado que esperabas? Argumenta tu respuesta.</h4>

<p>Sí, se observa una mejora importante al optimizar el tamaño de la red neuronal:</p>

<p><strong>Resultados con 5 Capas y 20 Neuronas:</strong> La precisión global es del 90%, pero varía significativamente entre clases. Algunas clases tienen un rendimiento bajo en comparación con otras.</p>

<p><strong>Resultados con la Mejor Configuración (50, 20):</strong> La precisión global mejora al 99%, con un rendimiento más consistente y alto en todas las clases. La mayoría de las clases tienen una precisión, recall y F1-score de 1.00.</p>

<p><strong>Argumento:</strong></p>
<ul>
  <li><strong>Mejora Significativa:</strong> La optimización lleva a una precisión general mucho mayor y una mayor consistencia en el rendimiento entre clases.</li>
  <li><strong>Explicación:</strong> La configuración optimizada puede capturar patrones más complejos en los datos, mejorando la capacidad del modelo para clasificar correctamente.</li>
  <li><strong>Resultado Esperado:</strong> La mejora es esperada, ya que una red con mayor capacidad puede aprender mejor de los datos, resultando en un mejor rendimiento global.</li>
</ul>

<h4>4.2 ¿Qué inconvenientes hay al encontrar el tamaño óptimo de la red? ¿Por qué?</h4>

<p><strong>Inconvenientes al Encontrar el Tamaño Óptimo de la Red:</strong></p>
<ul>
  <li><strong>Costo Computacional:</strong> Evaluar diferentes tamaños y configuraciones puede ser muy costoso en tiempo y recursos.</li>
  <li><strong>Sobreajuste:</strong> Redes demasiado grandes pueden ajustarse demasiado a los datos de entrenamiento, perdiendo capacidad de generalización.</li>
  <li><strong>Selección de Hiperparámetros:</strong> Además del tamaño, otros hiperparámetros también influyen en el rendimiento, complicando el ajuste.</li>
  <li><strong>Interacción Compleja:</strong> Cambios en el tamaño de la red afectan a otras partes, haciendo que la búsqueda del tamaño óptimo sea complicada.</li>
  <li><strong>Curva de Aprendizaje:</strong> La relación entre el tamaño de la red y el rendimiento no siempre es directa y puede requerir ajustes finos.</li>
</ul>